src/runtime/hexagon/hexagon_module.cc

```diff
@@ -0,0 +1,561 @@
+
+  std::vector<uint64_t> local_ss(size_ss / 8);
+  for (int i = 0; i != ndim; ++i) local_ss[i] = t->shape[i];
+  if (t->strides) {
+    for (int i = 0; i != ndim; ++i) local_ss[ndim + i] = t->strides[i];
+  }
+
+  hexagon_device_->CopyHostToDevice(remote, &local, sizeof local);
+  hexagon_device_->CopyHostToDevice(remote_ss, local_ss.data(), size_ss);
+  return remote;
+}
+
+hexagon::ArgLayout HexagonModuleNode::BuildArgLayout(const TVMArgs& As) const {
+  hexagon::ArgLayout Args;
+
+  for (unsigned i = 0, e = As.size(); i != e; ++i) {
+    const TVMArgValue& A = As[i];
+    unsigned TC = A.type_code();
+    switch (TC) {
+      // Treat all integers as 32-bit values.
+      case kDLInt:
+      case kDLUInt:
+        // KLUDGE: There is no distinction between 32- and 64-bit integer
+        // types, so there is no way to tell if the value being passed needs
+        // one or two registers. Assume that all integers are 32-bit, and
+        // simply abort if the actual value does not fit.
+        CHECK_EQ(static_cast<int64_t>(A), static_cast<int32_t>(A));
+        Args.Push(static_cast<int>(A));
+        break;
+      // 64-bit values
+      case kDLFloat:
+        Args.Push(static_cast<double>(A));
+        break;
+
+      case kTVMOpaqueHandle:
+      case kTVMNullptr:
+      case kTVMObjectHandle:
+      case kTVMModuleHandle:
+      case kTVMPackedFuncHandle:
+        Args.Push(static_cast<void*>(A));
+        break;
+
+      case kTVMNDArrayHandle:
+      case kTVMDLTensorHandle:
+        LOG(FATAL) << __func__ << ": cannot handle DLTensor*, code:" << TC;
+
+      default:
+        LOG(FATAL) << __func__ << ": unhandled type code" << TC;
+        break;
+    }
+  }
+
+  return Args;
+}
+
+Module HexagonModuleCreate(std::string data, std::string fmt,
+                           std::unordered_map<std::string, FunctionInfo> fmap,
+                           std::string asm_str, std::string obj_str,
+                           std::string ir_str, std::string bc_str,
+                           const std::set<std::string>& packed_c_abi) {
+  auto n = make_object<HexagonModuleNode>(data, fmt, fmap, asm_str, obj_str,
+                                          ir_str, bc_str, packed_c_abi);
+  return Module(n);
+}
+
+// Load module from file.
+Module HexagonModuleLoadFile(const std::string& file_name,
+                             const std::string& format) {
+  std::string data = file_name;
+  std::unordered_map<std::string, FunctionInfo> fmap;
+  std::string fmt = GetFileFormat(file_name, format);
+  std::string meta_file = GetMetaFilePath(file_name);
+  LoadMetaDataFromFile(meta_file, &fmap);
+
+  std::string empty;
+  // This passes {} as the set of packed C functions. Won't work for
+  // standalone functions on target.
+  return HexagonModuleCreate(data, fmt, fmap, empty, empty, empty, empty, {});
+}
+
+namespace hexagon {
+
+std::shared_ptr<Device> Device::Global() {
+  // Declare device constructors.
+#ifdef __ANDROID__
```

Comment:
If we want to support more Hexagon target in the future, we maybe have problem. For example, hexagon could run embed system of Ubuntu, which is official support of Hexagon SDK too. Then we can not use __ANDROID__ to distinguish target or simulator. One proposal is in our Hexagon.cmake, when we build for simulator, we define one macro like `-D__HEXAGON_SIM__=1` to indicate we are running on simulator or like `-D__HEXAGON_DEVICE__=1` to indicate we are running on real devices. This is one nit-pick comment, however, I wish we could consider it.

Reply:
We can probably do that, but I'd prefer to do it in another commit.